src/node/plugin.ts

```diff
@@ -1,92 +1,189 @@
-import { field, logger } from "@coder/logger"
-import { Express } from "express"
+import { Logger, field } from "@coder/logger"
+import * as express from "express"
 import * as fs from "fs"
 import * as path from "path"
-import * as util from "util"
-import { Args } from "./cli"
-import { paths } from "./util"
+import * as semver from "semver"
+import * as pluginapi from "../../typings/plugin"
+import { version } from "./constants"
+import * as util from "./util"
+const fsp = fs.promises
 
-/* eslint-disable @typescript-eslint/no-var-requires */
+interface Plugin extends pluginapi.Plugin {
+  /**
+   * These fields are populated from the plugin's package.json.
+   */
+  name: string
+  version: string
+  description: string
 
-export type Activate = (app: Express, args: Args) => void
-
-/**
- * Plugins must implement this interface.
- */
-export interface Plugin {
-  activate: Activate
+  /**
+   * path to the node module on the disk.
+   */
+  modulePath: string
 }
 
-/**
- * Intercept imports so we can inject code-server when the plugin tries to
- * import it.
- */
-const originalLoad = require("module")._load
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-require("module")._load = function (request: string, parent: object, isMain: boolean): any {
-  return originalLoad.apply(this, [request.replace(/^code-server/, path.resolve(__dirname, "../..")), parent, isMain])
+interface Application extends pluginapi.Application {
+  plugin: Plugin
 }
 
 /**
- * Load a plugin and run its activation function.
+ * PluginAPI implements the plugin API described in typings/plugin.d.ts
+ * Please see that file for details.
  */
-const loadPlugin = async (pluginPath: string, app: Express, args: Args): Promise<void> => {
-  try {
-    const plugin: Plugin = require(pluginPath)
-    plugin.activate(app, args)
-
-    const packageJson = require(path.join(pluginPath, "package.json"))
-    logger.debug(
-      "Loaded plugin",
-      field("name", packageJson.name || path.basename(pluginPath)),
-      field("path", pluginPath),
-      field("version", packageJson.version || "n/a"),
-    )
-  } catch (error) {
-    logger.error(error.message)
+export class PluginAPI {
+  private readonly plugins = new Array<Plugin>()
+  private readonly logger: Logger
+
+  public constructor(
+    logger: Logger,
```

Comment:
Is there an advantage to passing this in over `import`ing it?

Reply:
Yes, it  can be named for tests for example. i.e test specific logging at some point.

Can revert if you'd prefer importing.

Reply:
Nah, that's fair.